Different time controls per player. Allow user to set a different time control for player 1 and player 2.

James-Harris14 opened the pull request:
Hey,

I was looking through the reviews for the app on the play store and noticed a lot of requests for support to have a different time control for each player. I've built the feature to support this functionality. See https://github.com/ChessCom/android-chessclock/issues/5. I've added a checkbox to enable the feature in the same place as the full screen checkbox. I have added images of the feature in this description.

Let me know what you think and if you have any suggestions : )
![screenshot_20180101-110131](https://user-images.githubusercontent.com/16504199/34469016-810fac98-eee3-11e7-8285-cfc656d1d8ad.jpg)
![screenshot_20180101-110110](https://user-images.githubusercontent.com/16504199/34469017-826531a8-eee3-11e7-804d-6ffef75de09a.jpg)




Thanks,
James

autonomousapps commented:
Hi James! Thanks so much for the contribution, and sorry for the delay in responding to you. The team has been working hard on updating the main Chess app :)

I've talked to one of our product guys, and he had some comments on your implementation. I'm going to paste it below:

> It’s difficult to know from just the screenshots whether or not this is objectively good or bad; certainly, it’s quite different from the implementation in iOS, which *does* support time-based handicapping. In iOS, you can get there two ways:
> 1. Create a single named time-control with different initial time values for players 1 and 2. Note the Player tabs and the “Same as” switch here: https://i.imgur.com/yLbzvXx.jpg
> 2. You just adjust the player clocks on the fly from the clock interface itself.
> In the case of this submission, it seems like the notion is that you would never create a time control record with times for each player. Rather, the time controls are single-clock specific and chosen for each player. Technically, that’s valid, but given that 99% of times there will be no handicapping, it’s maybe a little much to force the user to select player time-controls discretely.
> Given that, I also find it strange that in the screenshot, the Player 1 list of controls isn’t exactly the same as the Player 2 list. I can’t imagine maintaining two separate lists of options for Player 1 and 2. That’s pretty laborious.

He then went on to show some screenshots from the iOS app to demonstrate how it was implemented there.

> * https://i.imgur.com/2lgBKXL.png : Clock face, not running. Note icons for Reset (which has a confirmation dialog), Settings, and Manual time adjustment for each side. (Time adjustment is just a conventional time-picker.)
> * https://i.imgur.com/5AA4BUH.png : Clock face, running. Note small integer at lower right of each clock. That’s a move counter.
> Note: To start the clock, the Black player (on whichever side) traditionally hits their button, starting their opponent's clock. This initial tap does not increment Black’s move counter.
> * https://i.imgur.com/raSxzVf.jpg : Main Settings screen, with menu of previously maintained time controls. (This shot shows the app in its default state, with the configuration / time controls you get, “out of the box.” “Edit” enables deletion or editing of time controls.
> * https://i.imgur.com/iwssm4F.png : Time control configuration (Player 1) and https://i.imgur.com/yLbzvXx.jpg (Player 2)
> * Add up to three separate stages to each time control: https://i.imgur.com/gCSdZIO.png -> https://i.imgur.com/rhYe2Bn.png 
> * All stages save the last (or only) require a # of moves that must be made in the configured time-span: https://i.imgur.com/zP2UNJk.png. By default, we assume 40 moves for Stage 1 (when Stage 2 is added) and 20 moves for Stage 2 (when Stage 3 is added). The final stage is always “Game in [time]” — i.e. ALL moves must be made before time runs out. No “moves” field here: https://i.imgur.com/kHg8YT8.png 
> * Interval definition = an amount of time for each move (always 5s by default) and an interval type:
>   * Fischer (default) - https://i.imgur.com/fyS1Ikd.png 
>   * Delay - https://i.imgur.com/B9wx4Hv.png
>   * Bronstein - https://i.imgur.com/sVkyLsr.png

If you can address this feedback and make it simpler for players to add different time controls, that would be great! Thank you.

James-Harris14 commented:
Hey @autonomousapps,

I've been at school for the past four months and wasn't able to work on this ticket. However, now that school is done for the term I've found some time to work on it. I've addressed your comments and made the look and feel similar to the iOS app.
![screenshot_20180422-140938](https://user-images.githubusercontent.com/16504199/39098337-362674ea-4637-11e8-8d6e-5bb00240a743.jpg)
![screenshot_20180422-141004](https://user-images.githubusercontent.com/16504199/39098339-38d76410-4637-11e8-925f-d75730ebad28.jpg)


Let me know what you think.

Best,
James

autonomousapps commented:
James, thanks! I'll take a look as soon as I can. We really appreciate the contribution.

autonomousapps commented:
@sklinefelter would you like to give this a look?

sklinefelter commented:
I would be more than happy to

sklinefelter commented on app/src/main/java/com/chess/clock/activities/SettingsActivity.java:
```diff
@@ -77,6 +83,11 @@ public void onServiceDisconnected(ComponentName arg0) {
+    /**
+     * BottomNavigationTab
+     */
+    BottomNavigationView mBottomNavigationView;
```
Can this be private instead of package-private?

sklinefelter commented on app/src/main/java/com/chess/clock/activities/SettingsActivity.java:
```diff
@@ -308,4 +321,18 @@ public void saveTimeControl() {
+    }
+
+    public void setBottomNavigationViewVisibility(int v) {
+        mBottomNavigationView.setVisibility(v);
```
With a null check in `setNavigationOnItemSelectedListener` I feel it would also be appropriate to have a null check here in this method.

sklinefelter commented on app/src/main/java/com/chess/clock/activities/SettingsActivity.java:
```diff
@@ -308,4 +321,18 @@ public void saveTimeControl() {
+    }
+
+    public void setBottomNavigationViewSelected(int id) {
+        mBottomNavigationView.setSelectedItemId(id);
```
For consistency if nothing else, I would also prefer a null check here as well.

sklinefelter commented on app/src/main/java/com/chess/clock/adapters/TimeControlAdapter.java:
```diff
@@ -6,17 +6,17 @@
 
 
-public class TimeControlAdapter extends ArrayAdapter<TimeControl> {
+public class TimeControlAdapter extends ArrayAdapter<TimeControlWrapper> {
```
I'm intrigued by this new wrapper class...I will look for it later on in this PR

sklinefelter commented on app/src/main/java/com/chess/clock/engine/TimeControlManager.java:
```diff
@@ -172,17 +172,22 @@ public void prepareNewEditableTimeControl() {
+        try {
+            mEditableTimeControl = new TimeControlWrapper(blank, (TimeControl) blank.clone());
+        } catch(CloneNotSupportedException e) {
+            e.printStackTrace();
```
Hhmm...I'm torn between if we want to just log or log and throw an `IllegalStateException` so that we know there is trouble here...

I understand it "impossible" for this to occur in theory, but I'm not very trusting :P

James-Harris14 replied:
Fixed in 0faf160

sklinefelter commented on app/src/main/java/com/chess/clock/engine/TimeControlManager.java:
```diff
@@ -211,17 +216,12 @@ public void setEditableTimeControlCheckIndex(int idx) {
-                e.printStackTrace();
-            }
+            TimeControlWrapper original = mTimeControls.get(position);
+            return new TimeControlWrapper(original.getTimeControlPlayerOne(), original.getTimeControlPlayerTwo());
```
Maybe we should use a Factory pattern here instead of `newing`?

sklinefelter commented on app/src/main/java/com/chess/clock/engine/TimeControlParser.java:
```diff
@@ -165,22 +175,32 @@ public static int getLastTimeControlCheckIndex(Context context) {
                 JSONObject timeIncrementJSONOBject = timeControlJSON.getJSONObject("timeincrement");
                 JSONArray timeControlStagesJSON = timeControlJSON.getJSONArray("stages");
 
+                if (timeControlJSON.has("timeincrement2") && timeControlJSON.has("stages2")) {
```
Maybe we should pull these strings into constants.  Also, would there be better names than just adding "2"?

sklinefelter commented on app/src/main/java/com/chess/clock/engine/TimeControlParser.java:
```diff
@@ -198,42 +218,48 @@ public static int getLastTimeControlCheckIndex(Context context) {
-        timeControls.add(blitzTimeControl);
-        timeControls.add(fischerRapid);
-        timeControls.add(tournamentTC);
+        timeControls.add(fischerBlitzTCWrapper);
```
I'm not a fan of the fact that some of these use `TC` while some use `TimeControl` which I realize is not related to your work and this PR...just something I stumbled upon :)

James-Harris14 replied:
The naming scheme is fixed in 0faf160. All default time controls are named `TC`

sklinefelter commented on app/src/main/java/com/chess/clock/fragments/TimeControlFragment.java:
```diff
@@ -48,7 +70,9 @@
-    private TimeControl mTimeControl;
+    private TimeControlWrapper mTimeControlWrapper;
+    private TimeControl mSelectedTimeControl;
+    private boolean mplayerOneSelected = false;
```
The P should be capitalized here

sklinefelter commented on app/src/main/java/com/chess/clock/fragments/TimeControlFragment.java:
```diff
@@ -117,7 +144,8 @@ public void onAttach(Activity activity) {
             // Fetch current TimeControl object
-            mTimeControl = mListener.getEditableTimeControl();
+            mTimeControlWrapper = mListener.getEditableTimeControl();
+            Log.e("", mTimeControlWrapper.toString());
```
Let's say something other than an empty string here please...actually...does this need to be logged, and does this need to be logged as an error?  I'm slightly confused about this line now.

James-Harris14 replied:
This was for my debugging purposes. I'll remove this.

sklinefelter commented on app/src/main/java/com/chess/clock/fragments/TimeControlFragment.java:
```diff
@@ -150,38 +179,61 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
 
-                if (mTimeControl.getName() != null && !mTimeControl.getName().equals("")) {
-                    mTimeControlNameEditText.setText(mTimeControl.getName());
+                TimeControl tc = mTimeControlWrapper.getTimeControlPlayerOne();
```
Is getting player one's time controls always going to be correct?

James-Harris14 replied:
I figured that before when we were using one time control we used it as the source of truth. So with the addition of the time control wrapper, the player one time control will be the source of truth when the fragment is first created.

sklinefelter commented on app/src/main/java/com/chess/clock/fragments/TimeControlFragment.java:
```diff
@@ -194,6 +246,9 @@ public void onClick(View v) {
             }
         }
 
+        ((SettingsActivity) getActivity()).setBottomNavigationViewSelected(R.id.nav_player1);
```
Hhmm...I wonder if there's a way we can get rid of that cast...

James-Harris14 commented:
@sklinefelter addressed almost all of your comments in 0faf160

sklinefelter commented:
Hey James...thanks a lot for taking the time to address my feedback.  I have some comments from Michael who is one of our product guys:

> Well, that’s actually pretty great! There are a few little things that should be fixed, but I really like it.
> 
> 1. Minor. The Moves counter appears only for multi-stage time controls. It should always be there, as it’s interesting to know how many moves were played in the game. For the same reason, in iOS we do not restart the move counter at the beginning of each stage. (But whether or not it restarts is less of a big deal. People can do the math.)
> 
> 2. This one is a legit bug: Let’s say you have three stages. Stage 1 is 10 moves in 20 minutes; Stage 2 is 5 moves in 10 minutes, and Stage 3 is Game in 5 mins (just for the sake of argument). When Stage 2 begins (after the player makes her 10th move), her clock is awarded +20 minutes (the duration of the _previous_ stage), not +10 mins, as it should be. Likewise at the start of the third stage, where the player gets +10 mins instead of the correct +5.
> 
> 3. (Minor) When you add a stage, the # of moves for the previous stage defaults to 1, which is _always_ going to wrong. This should be 20, which is more conventional. (I’m talking about editing the Time Control here.) 
> 4. Lastly, there is some wonkiness I think with the switch that sets Player 2's settings to = that of Player 1. First off, when I create a new time control, that should default to ON, even if I was using an “odds” time control just before that. If it IS different already, and you turn the switch to ON, you have to save and exit and reopen the config to see that the settings for Player 2 have actually been updated to match Player 1's. Not the end of the world, but a little twitchy.
> Finally, the time odds functionality worked great!
> I think that’s all I’ve got.

Let me know if you need any clarification or when you're ready for another code review pass at this PR.  Appreciate your work!

James-Harris14 commented:
Hey @sklinefelter,

I've address the above comments:
1. 8900baa -> The moves count is now always shown, but it is reset when a new stage is activated. Let me know if this is acceptable.

2. I tested this and it seems like the functionality is actually already implemented. I couldn't find anything wrong with it.

3. b4e376f -> The number of moves in a new stage is now defaulted to 20

4. d5335ad -> As per requested. The default state of the switch when creating a time control is now on. Also I fixed the bug so that the user must now save and exit when modifying an existing time control. Finally, I have added functionality for a "real time" update of the player 2 time control when the switch is set to on. As per requested, you will not have to back out to view the changes to player two's time control.

As always keep the feedback coming :) It's been an honour to collab with you guys so far!!

sklinefelter commented:
@James-Harris14 I think we're getting close to accepting this PR!  Glad you've enjoyed collaborating as I have with you.

As for the second issue...I'm going to walk through a small scenario assuming there are three stages:

Stage 1 - x moves in 20 mins
Stage 2 - y moves in 10 mins
Stage 3 - z moves in 5 mins

Michael noted that:

>"[the] clock is awarded +20 minutes (the duration of the _previous_ stage), not +10 mins, as it should be"

Michael's point is that there is an error when it comes to the amount of time awarded when Stage 2 and Stage 3 begins.  Meaning that when Stage 2 starts, it's showing Stage 1's time of 20 mins, and when Stage 3 beings it's showing Stage 2's time of 10 mins.  Does that make more sense?  Hopefully, I can translate enough :)  

I am uncertain if this is existing without doing a deeper dive into the code base.  Please feel free to let me know if this is existing or new based on these changes and we can go from there.

sklinefelter commented:
Hey @James-Harris14!  I'm sure you are busy, but just wanted to check in since we would love to merge this awesome functionality!  Please let me know if you need any clarification on the last requested change or anything from our end.

James-Harris14 commented:
Hey @sklinefelter,

Sorry for the delay. I've been staying late at my current internship as work has recently been hectic. I've attached a video and verified the behaviour of the mentioned comment. It seems to be working as expected. Notice how when stage 2 is activated both clocks receive +10 minutes, and when entering stage 3 both players receive +5 minutes.


![stages](https://user-images.githubusercontent.com/16504199/41207119-ef153f0a-6cdd-11e8-9d4b-77d08876067a.gif)

sklinefelter commented:
@James-Harris14 Thank you for the upload!  I too found the same behavior when I poked around to check.  I will discuss more with Michael (our product guy) to see if there are any steps missing to reproduce what he believed was an issue.  I have also sent him a new APK with the updates you have made so I'm hopeful we can merge this sometime next week.

sklinefelter commented:
Hey @James-Harris14 !  Firstly, I want to apologize for the extremely long time it has taken for me to provide feedback.  We have a lot of exciting things going on at Chess.com on the Android front and we've all been super busy.  

I have two minor discussion points.

1. Can you explain to me the changes for the dots for the new time controls?  I believe they are now visible due to changes in these two files:

- https://github.com/James-Harris14/android-chessclock/blob/d5335adfe83b95859dfda6f0734e96e531222243/app/src/main/res/layout-land/activity_clock_timers.xml
- https://github.com/James-Harris14/android-chessclock/blob/d5335adfe83b95859dfda6f0734e96e531222243/app/src/main/res/layout-land/activity_clock_timers_reversed.xml

2.  Our product person (Michael) feels that we should avoid resetting the move cound to zero between each stage, is that easily doable?

Thanks again for collaborating and for the effort and patience.  It's really neat to see this improvement in ChessClock and we're anxious to merge it!

James-Harris14 commented:
@sklinefelter I will look into making these changes and get back to you ASAP

James-Harris14 commented:
Hey @sklinefelter I've had some time to look into the two points of feedback.

1. The visibility of the dots is the same as before this change. When there are stages in the time clock, the the dots will represent the total number of stages and the highlighted dot will represent the stage that the player is currently on. This can be easily removed if necessary.

2. It should be fairly easy to not reset the move counter to zero between each stage.

greatwolf commented:
What's the current status of this PR? I was hoping this would be merged in by now.

sklinefelter commented:
Hey there!  I apologize for the delay.  We've been extremely busy with awesome changes to the Chess.com Android app.  

Unfortunately, I don't feel comfortable merging this PR until the last two points of feedback are addressed.  We would welcome any assistance.  With all of the exciting other work, Chess Clock has just not been as high of a priority for our small team, but we still love it.

Also, thank you, James, for working with us and your effort/patience on this PR.

James-Harris14 commented:
Hi @sklinefelter and @greatwolf, I'm extremely sorry I've neglected this PR I've been extremely busy over the past year with school and internships. I will take another look at the comments and address everything by the end of the week.

sklinefelter commented:
@James-Harris14 I understand that life gets in the way sometimes, no worries. Please ping me when this is ready for final review and best of luck with school and internships and everything.

James-Harris14 commented:
Hi again @sklinefelter,

I've addressed feedback #1 about the stage dots in 7c2f73c, and I have addressed feedback #2 in 985902a so the move counter is not being reset to zero between stages now.

I believe with these last two changes the PR should be in a good place to be merged in! It's been really awesome working with you 😄 Once again sorry for the delays along the way!

Very best,
James

sklinefelter commented:
@James-Harris14 The last two code updates look reasonable and like they should make this PR mergable.  I will test some tomorrow and then take next steps as necessary to start a release.  Thank you again for your contribution.
